File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonVoteSerializer.java
package com.github.dedis.popstellar.model.network.serializer;

import com.github.dedis.popstellar.model.network.method.message.data.election.*;
import com.google.gson.*;

import java.lang.reflect.Type;

/** Simple serializer to convert json elements to Vote types and vice-versa */
public class JsonVoteSerializer implements JsonSerializer<Vote>, JsonDeserializer<Vote> {

  @Override
  public JsonElement serialize(Vote src, Type typeOfSrc, JsonSerializationContext context) {
    // make sure the actual class of the vote is used for serialization.
    // Even if the asked type is Vote
    return context.serialize(src);
  }

  @Override
  public Vote deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    IntermediateVote inter = context.deserialize(json, IntermediateVote.class);
    return inter.convert();
  }

  @SuppressWarnings("unused")
  private static class IntermediateVote {
    private String id;
    private String question;
    private JsonPrimitive vote;

    public Vote convert() {
      if (vote.isString()) {
        return new EncryptedVote(id, question, vote.getAsString());
      } else if (vote.isNumber()) {
        return new PlainVote(id, question, vote.getAsInt());
      } else {
        throw new JsonParseException("Unknown vote type");
      }
    }
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/data/JsonVoteSerializer.kt
package com.github.dedis.popstellar.model.network.serializer.data

import com.github.dedis.popstellar.model.network.method.message.data.election.EncryptedVote
import com.github.dedis.popstellar.model.network.method.message.data.election.PlainVote
import com.github.dedis.popstellar.model.network.method.message.data.election.Vote
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type

/** Simple serializer to convert json elements to Vote types and vice-versa  */
class JsonVoteSerializer : JsonSerializer<Vote?>, JsonDeserializer<Vote> {
    override fun serialize(src: Vote?, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        // make sure the actual class of the vote is used for serialization.
        // Even if the asked type is Vote
        return context.serialize(src)
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Vote {
        val inter = context.deserialize<IntermediateVote>(json, IntermediateVote::class.java)
        return inter.convert()
    }

    @Suppress("unused")
    private class IntermediateVote {
        private val id: String? = null
        private val question: String? = null
        private val vote: JsonPrimitive? = null
        fun convert(): Vote {
            return if (vote!!.isString) {
                EncryptedVote(id, question, vote.asString)
            } else if (vote.isNumber) {
                PlainVote(id, question, vote.asInt)
            } else {
                throw JsonParseException("Unknown vote type")
            }
        }
    }
}